### PR TITLE
Merge 3x to stable

### PR DESCRIPTION
**Pull request by MikaelSmith**

### CODE REVIEW

File: lib/puppet/pops/validation/checker4_0.rb [hunk 1]
```diff
@@ -317,6 +323,12 @@ def check_FunctionDefinition(o)
     #  but not be special in any other way
     #
     check_NamedDefinition(o)
+
+    # This is perhaps not ideal but it's very difficult to pass a ReservedWord through
+    # the mechanism that creates qualified names (namestack, namepop etc.)
+    if FUTURE_RESERVED_WORDS[o.name]
+      acceptor.accept(Issues::FUTURE_RESERVED_WORD, o, {:word => o.name})
```
Comment (MikaelSmith) [inline lib/puppet/pops/validation/checker4_0.rb]: @thallgren @hlindberg can you check this spot in particular? The structure of check_FunctionDefinition changed a bit, so I'm not sure whether this is right or if the fix for [PUP-5036](https://tickets.puppetlabs.com/browse/PUP-5036) even needs to be merged up.

**Comment by MikaelSmith** [comment]
I regenerated eparser.rb using `rake gen_eparser` after resolving everything else.

**Comment by MikaelSmith** [comment]
@kylog @joshcooper do either of you have background on what needs to be done for a merge?

**Comment by kylog** [comment]
@MikaelSmith we do want PUP-5036 merged up, but I'd let Henrik or Thomas have a crack at sorting the mergeup.

**Comment by MikaelSmith** [comment]
@hlindberg @thallgren if you want to take care of merging https://github.com/MikaelSmith/puppet/commit/5d85dc56822d9535ba5f4afa446703371892b915 while I'm asleep, I can take care of the later commits tomorrow.

**Comment by thallgren** [comment]
Merge to stable at https://github.com/puppetlabs/puppet/commit/813744d9fb2f1fab178e4ccfefdb48d0d5136872 and to master at https://github.com/puppetlabs/puppet/commit/90a0b4a07d5b60bf8b95df6e20eaa72f362ca446.

**Comment by MikaelSmith** [comment]
That didn't merge PR #4167 correctly :( trying to fix it up.

**Comment by thallgren** [comment]
Oops. Sorry about that. There was a whole slew of MacOS/X related stuff removed from stable due to no longer supporting older releases. All of them were in conflict and from the looks of it, PR https://github.com/puppetlabs/puppet/pull/4167 was about changes in code that had been removed on stable. My mistake. Will be more careful next time.

**Pull request closed without merging**